### ISSUE TITLE
Timber::get_term() API: clarify taxonomy requirements

### ISSUE THREAD

**Issue by acobster** [timber/timber#2087]
## Is your feature request related to a problem? Please describe.

Per #2073, we're moving away from direct instantiation of Terms in favor of `Timber::get_term()`. The problem here is that simply changing from `new Timber\Term($term_id)` to `Timber::get_term($term_id)` will break existing code in most cases, because the current `Timber::get_term()` implementation uses a default `$taxonomy` (second arg) of `post_tag`.

The following tests in `tests/test-timber-term.php` fail if the constructor call in each is changed to a direct `Timber::get_term(...)` call _without also specifying a `$taxonomy`_:

* `testConstructor`
* `testTermChildren`
* `testTermWithNativeMeta`
* `testTermWithNativeMetaFalse`
* `testTermWithNativeMetaNotExisting`
* `testTermEditLink`

## Describe the solution you’d like

It would be better if the `$taxonomy` arg to `Timber::get_term()` defaulted to the empty string. This would be more in line with the [core API](https://developer.wordpress.org/reference/functions/get_term/#parameters), and I believe it does not constitute a breaking change to existing calls to `Timber::get_term()`.

If possible, `Timber::get_term()` should also support things that `Term::__construct()` currently supports, i.e.:

- [x] ~Passing a slug as first param~
- [x] Rather than passing a slug directly to `::get_term()`, for consistency with the new Post/User APIs, we should implement a `::get_term_by()` method. #2186 
- [x] Passing an object as first param

## Describe alternatives you’ve considered

A sub-optimal solution (which is being incorporated into the refactor on #2085) is to require a second `$taxonomy` param in every case. This would likely mean considerably more work for upgrading downstream code, over and above the work required to migrate away from direct instantiation.

## Additional context

#2085 introduces a quick way to run tests that call, or did at one point call, the `Timber\Term` constructor:

```
phpunit --group called-term-constructor
```

Note at time of writing this is not on `2.x` yet, but hopefully will be soon!

**Comment by acobster** [comment]
@jarednova - would be good to get your thoughts on why `"post_tag"` was originally chosen as the default instead of `""`. Thanks!

**Comment by jarednova** [comment]
@acobster I believe this goes back to waaaaaaay long ago days (~2012) when WP Core assumed `post_tag` as the "default" tag in a number of situations. Things have come a long way since then, and based on the current docs, it looks like that's no longer the case (or at least, that assumption has been deprecated)

**Comment by gchtr** [comment]
> It would be better if the `$taxonomy` arg to `Timber::get_term()` defaulted to the empty string. This would be more in line with the [core API](https://developer.wordpress.org/reference/functions/get_term/#parameters), and I believe it does not constitute a breaking change to existing calls to `Timber::get_term()`.

I think it’s better if we can match the core API. And I think we can probably do better than WordPress Core and describe in more detail what the default argument `""` does.

As far as I can tell from looking at the Core code, the result shouldn’t change for most existing calls to `get_term()`. It could only matter if a term is shared between different taxonomies. For example: If the term is only used in a taxonomy other than `post_tag`, you could still get a result. Is that right?

> If possible, `Timber::get_term()` should also support things that `Term::__construct()` currently supports

Yes! Would you add «Passing a term ID» for completeness as well?

**Comment by acobster** [comment]
> Yes! Would you add «Passing a term ID» for completeness as well?

@gchtr I was mostly thinking of passing zero args, and getting `term_id` from the current query.

> For example: If the term is only used in a taxonomy other than post_tag, you could still get a result. Is that right?

Sounds right to me. The tests should probably cover what happens when you share a Term between taxonomies. I didn't even know that was a thing; seems like kind of a nonsensical use-case to me. Do you have an example of a setup like that?

**Comment by gchtr** [comment]
> @gchtr I was mostly thinking of passing zero args, and getting `term_id` from the current query.

👍

> Sounds right to me. The tests should probably cover what happens when you share a Term between taxonomies.

👍

> I didn't even know that was a thing; seems like kind of a nonsensical use-case to me. Do you have an example of a setup like that?

I never had a use case for this and can’t think of one, either.

**Comment by acobster** [comment]
So, we definitely don't explicitly support sharing terms between taxonomies, as demonstrated by the fact that when we don't use the `$taxonomy` arg _at all_ (see 3cc9c97), all the term API tests still pass. I'm asking around on #wordpress IRC and also [Stack Exchange](https://wordpress.stackexchange.com/questions/357211/what-is-the-use-case-for-sharing-a-term-between-multiple-taxonomies), but unless I hear a compelling use-case for sharing terms, I'm leaning toward explicitly not supporting it. Seems like a bad idea to begin with.

**Comment by jarednova** [comment]
@acobster I'm in agreement that the sharing of terms is one of those hypothetical things I've never seen a legit use case for. Even when programmatically creating two identically named terms in diff't taxonomies. WP (as of 5.3.2) will create two separate terms, as demonstrated by this test:

```php
function testSharedTerm() {
	register_taxonomy('nfl', 'post');
	register_taxonomy('mlb', 'post');
	$cards_nfl_id = $this->factory->term->create( array('name' => 'Cardinals', 'taxonomy' => 'nfl') );
	$cards_mlb_id = $this->factory->term->create( array('name' => 'Cardinals', 'taxonomy' => 'mlb') );
	$term_ids = array($cards_mlb_id, $cards_nfl_id);
	$terms = Timber::get_terms($term_ids);
	$this->assertEquals('cardinals', $terms[0]->slug);
	$this->assertEquals('cardinals', $terms[1]->slug);
	$this->assertNotEquals($terms[0]->ID, $terms[1]->ID);
}

**Comment by acobster** [comment]
Eureka! Someone on IRC dredged up a [blog post](https://make.wordpress.org/core/2015/02/16/taxonomy-term-splitting-in-4-2-a-developer-guide/) from 2015 about this. Basically, since 4.3 WP has made sure on term update that any shared terms get split up into distinct ones.

> The vast majority of terms are not shared between taxonomies; shared terms themselves are an odd edge case.

So the only scenario I can imagine shared terms taking effect at all for a given site is if `wp_update_term()` has literally never been called since that site was running on 4.2. Passing the non-default, non-empty arg is definitely more effort for very little benefit.

To sum up, this means we can safely drop the `$taxonomy` arg altogether.

**Comment by jarednova** [comment]
bye bye!
